### ISSUE TITLE
messages are no longer retained with 2.0.16 (or 2.0.17)

### ISSUE THREAD

**Issue by BubuOT** [eclipse/mosquitto#2887]
Since upgrading to mosquitto 2.0.16 (and later 2.0.17) (via docker image) retained messages received by the broker are no longer handled as retained and not delivered to new subscribers.

Downgrading to 2.0.15 makes this work as expected again.

**Comment by stigvi** [comment]
> Downgrading to 2.0.15 makes this work as expected again.

Yes, I did the same after a lot of troubles with no retained messages with 2.0.17

**Comment by BubuOT** [comment]
@stigvi Good to know we are at least not alone with this problem. Did you find a way to reproduce this problem somehow?

**Comment by pat1** [comment]
duplicated of #2806 and #2618

**Comment by pat1** [comment]
related to #2785

**Comment by BubuOT** [comment]
@pat1 
> duplicated of https://github.com/eclipse/mosquitto/issues/2806 and https://github.com/eclipse/mosquitto/issues/2618

I'm not sure it is? We do *not* experience this problem here with 2.0.15, which is explicitly mentioned in the other issue(s) as affected.

**Comment by ralight** [comment]
@BubuOT I've been unable to reproduce this so far. Are you able to share any more details about the setup and what sorts of topics you are using?

**Comment by clintkev251** [comment]
I've also seen this. Specifically I'm using mosquitto with Home Assistant and Zigbee2MQTT in this case. Previously, when Home Assistant was restarted, it would read device configurations set as retained messages from Zigbee2MQTT automatically. However on 2.0.17, it would seemingly never be able to receive these retained messages and as a result devices remained unavailable until a restart of Zigbee2MQTT to force it to republish all of the discovery topics at which point they would be successfully received by Home Assistant

I confirmed that rolling back to 2.0.15, this began working as expected again. Let me know if there's any other information that would be helpful in reproducing

**Comment by simonclayton** [comment]
We have this exact issue when upgrading mosquitto 2.0.15 to 2.0.18 on docker. I can replicate the issue with data "missing" on 2.0.16, 2.0.17 and 2.0.18 but switching back to the 2.0.15 docker image just by editing the docker-compose.yml file and restarting immediately "fixes" everything without needing to republish any data.

In case it helps, we only start to see the issue with increasing volumes of data. We have 2 top level topics with a number of nested topics and at ~30k topics published on 2.0.15 or 2.0.18, connecting with MQTT Explorer shows ~1400 topics as the complete data set. 

Testing with only one top level topic with ~1100 subtopics, we get all of the data sent to the client correctly. It is only when we add a 2nd top level topic with ~28k subtopics that we start to see issues.

For reference - our topic structure looks like this

--- topic1 (~1100 subtopics in total)
 |       |- sub1
 |       |     |- sub-subtopics1..10
 |       |     |                |- data topics
 |       | 
 |       |- sub2
 |       |     |- sub-subtopics1..2
 |       |     |                |- data topics
 |       | 
 |- topic2 (~28k subtopics in total)
 |       |- sub1
 |       |     |- sub-subtopics1..10
 |       |     |                |- data topics
 |       | 
 |       |- sub2
 |       |     |- sub-subtopics1..2
 |       |     |                |- data topics

I have a test server that only has dockerized mosquitto on and no personal data that I can let a project dev have root access to if you want to see this happening.

**Comment by BubuOT** [comment]
FWIW, we also have a lot of retained topics (~10k) in production where we saw this issue. This might explain why I've never been able to reproduce this in any test setup :thinking:.

**Comment by ZetaWaves** [comment]
I am having the same issue. All devices come up as "Unavailable" on Home Assistant after an HA reboot. 
Is this being worked on? This is a critical bug as my entire network is down with no fix.

**Comment by clintkev251** [comment]
Well you can temporarily fix it very easily by reverting to 2.0.15 where retained topics are working as expected

**Comment by ZetaWaves** [comment]
> Well you can temporarily fix it very easily by reverting to 2.0.15 where retained topics are working as expected

Not possible. I’m using MQTT HA Addon. No way to go back in version unfortunately.

**Comment by Soukyuu** [comment]
I just restored to HA addon version 6.2.1, which still has mosquito 2.0.15 - backups are a life saver.

**Comment by ZetaWaves** [comment]
> I just restored to HA addon version 6.2.1, which still has mosquito 2.0.15 - backups are a life saver.

I'm happy for you, but unfortunately, I wasn't aware of when the problem started until I had to reboot HA.
So all my backups are useless because I've made changes since then and I can't go back especially since I'm not sure when I upgraded the MQTT broker add-on. So yeah - this completely breaks using any MQTT device for me. Network has been down for some time. I will likely drop MQTT at this point because the dev's don't seem to be making this a priority fix.

**Comment by Soukyuu** [comment]
@ZetaWaves: HA does partial backups before every update by default, are you sure you hsve none? Or jave you disabled that?

**Comment by WhimsySpoon** [comment]
I have the same issue with HA Core 2023.10.1, the HA Add-on v6.3.1 (2.0.17) and Zigbee2MQTT. Several devices became Unknown, even though they had a state in Z2M.

Restoring my add-on back to the version with 2.0.15 has resolved the issue. I have ~1400 topics.

**Comment by wardwolfram** [comment]
> I just restored to HA addon version 6.2.1, which still has mosquito 2.0.15 - backups are a life saver.

I issues a partial restore to MQTT broker 6.2.1.  After the restore, I received the following message:
![image](https://github.com/eclipse/mosquitto/assets/30361932/542b534b-1408-42fc-b398-a686480c31cb)


How did you restore to 6.2.1 successfully?  Thanks much in advance.

**Comment by tomlut** [comment]
That's something you should ask Home Assistant, not Mosqitto.

**Comment by wardwolfram** [comment]
> That's something you should ask Home Assistant, not Mosqitto.

Yes, restoring Mosqitto from HA.

**Comment by wardwolfram** [comment]
> > I just restored to HA addon version 6.2.1, which still has mosquito 2.0.15 - backups are a life saver.
> 
> I issues a partial restore to MQTT broker 6.2.1. After the restore, I received the following message: ![image](https://user-images.githubusercontent.com/30361932/275282741-542b534b-1408-42fc-b398-a686480c31cb.png)
> 
> How did you restore to 6.2.1 successfully? Thanks much in advance.

Solved. 

1. Stopped the Zigbee2MQTT add-on service
2. Stopped the Mosquitto Broker add-on service
3. Uninstalled the Mosquitto Broker add-on
4. HA partial restore of the Mosquitto Broker add-on v6.2.1
5. Started the 2 services up successfully.

I agree with Soukyuu... **backups are indeed a life saver!**

**Comment by Alphaemef** [comment]
anyone know if this has been fixed in the recent updates or if a fix is coming ? I am still sitting on 2.0.16 for the same reason.

**Comment by tomlut** [comment]
Yes there is a bugfix for this included. See:

https://mosquitto.org/blog/2023/09/version-2-0-18-released/

**Comment by Alphaemef** [comment]
Aha so this was actually #2893 Perfect, thanks!

**Comment by Alphaemef** [comment]
sigh nope... when I log on with MQTT explorer, none of the messages are retained.

**Comment by ZetaWaves** [comment]
It is not fixed in 2.0.18.

Still broken. Just tested.

**Comment by Alphaemef** [comment]
It may be an issue with the MQTT broker addon for HA. I have around 30 clients, each with around 100 entities. And not a single message is retained, despite several reboots etc. Retained messages however work with version 2.0.15 (Addon version 6.21). 

Anyone know who maintains the addon ?

**Comment by tomlut** [comment]
Frenk. You can have a chat on discord here: https://discord.gg/jBwTDzJ4sk

 I'm using the HA addon but don't have anywhere near that many topics.

**Comment by Alphaemef** [comment]
From what I can gather, its primarily hitting users with a lot of topics (I guess 5550 topics is high). Its so rarely reported though, I don't even know where to look. Will check out the discord.

**Comment by euggersh** [comment]
If you find out anything interesting on Discord, a post here would be very much appreciated.

**Comment by Alphaemef** [comment]
Tagged French. Will see if he has ny comments or guidance. Might just be time to host the MQTT broker outside HA, which I really was hoping to avoid.

**Comment by clintkev251** [comment]
I don't think it's anything specific to hosting the broker in HA, it's the same container regardless of where you're hosting it. I run mine completely independently and I'm still impacted by this

**Comment by BubuOT** [comment]
2.0.18 doesn't fix this, no.

**Comment by Alphaemef** [comment]
> Just adding that I ran into this problem running 2.0.18 and downgrading to 2.0.15 resolved the issue. Basically the broker just won't retain messages. I even manually try to set a retained message using MQTTExplorer and even that fails to retain. All good on 2.0.15

Yeah its likely the same issue. I totally get that this isn't necessarily fixed immediately. But it sure would be nice if it at least could get acknowledged as an actual issue.

**Comment by digiblur** [comment]
I can also confirm this issue.  Rolled back my container tag to 2.0.15 and all is well again with Zigbee2MQTT and HA.

**Comment by MnM001** [comment]
Not worried about this in version 2.0.15?

![image](https://github.com/eclipse/mosquitto/assets/15014858/d14f0e42-24b6-47dc-8772-fb31545e7a6f)

**Comment by stigvi** [comment]
> Not worried about this in version 2.0.15?

Why do you ask? What alternatives are there?

**Comment by MnM001** [comment]
No idea... new 2.0.15 image with vulnerabilities fixed?

I mean if you thought that it was bad when devices were not working, I really think it will way worst than that if someone manages to exploit these vulnerabilities.
And as the mqtt owner is aware that any versions after 2.0.15 have issues maybe they might have a duty of care to build a new 2.0.15 version without vulnerabilities?

**Comment by Alk3m1st** [comment]
Hey all just to say I was having these issues also with my setup. Tried a lot of things but in the end it was a Home Assistant Core and Operating System update that fixed it for me. I was going to do an add-on restore but luckily didn't need to.
Running on Core 2024.5.5 and OS 12.3.
Thought I'd post here in case anyone can't restore or would rather not do so. Hopefully this resolved it for others too.

**Comment by digiblur** [comment]
This is impacting people that don't even run HAOS.

**Comment by SVH-Powel** [comment]
Can this be solved by inceasing max_queued_messages?

**Comment by WhimsySpoon** [comment]
Update: installed
HA Core: restarted

Result: So far, so good.

The previous version with the issue resulted in devices showing as unavailable, whereas everything has come online as expected.

**Comment by SVH-Powel** [comment]
So far, so good here, too. I think I stay on 2.0.18. Default value for max_queued_messages is 1000 and I have set it to 8192. (0 is unlimited)

I have restarted my system several times and there is no problem with retained messages.

**Comment by ckrey** [comment]
> Ok, thank you. I will keep looking, and if you find a way to reproduce it please let me know.

There is a way to reproduce:
I am using an M1 MacBook Pro running both broker and client 2.0.18. The problem might occur earlier/later depending on the speed/memory/operating system

a) Setup mosquitto broker with `max_queued_messages 1000` and `log_dest file ...`
b) populate mosquitto with 100.000 retained messages. The command line below is not fast, but does the job:
```
for i in {0..99999}; do echo $i; mosquitto_pub -t "retained/$i" -m $i -r -q 0 ; done
```
c) subscribe to the broker to retrieve all the retained message. Make the subscriber "slow" by using debug and json output:
```
mosquitto_sub -t '#' -q 0 -d -v -F %j
```
As a result, only part of the messages will be retrieved.

d) The cause of the problem is in `mosquitto.log`
```
2024-06-27T17:32:17: Outgoing messages are being dropped for client auto-F28ED67E-3E1B-1C64-D868-F6C94FCF259E.
```
This behaviour was introduced by fixing CVE-2024-28366 in 2.0.16 in `lib/packet_mosq.c` https://github.com/eclipse/mosquitto/commit/6113eac95a9df634fbc858be542c4a0456bfe7b9

Not sure if the change was necessary to fix the CVE. Especially because `max_queued_messages` should only be valid for QoS1 or QoS2 messages.

To avoid the problem raise `max_queued_messages` to more than your maximum number of retained messages you intend to retrieve.
 
Using QoS1 or QoS2 for retained messages does not help as the  `max_inflight_messages 20` parameter does not seem to be respected in case of retained messages delivery.

**Comment by BubuOT** [comment]
@ralight now that there's a reproducer, could you maybe take another look at this issue?

**Comment by ralight** [comment]
Yes. It's more of an annoying problem to solve than it may appear though :(

**Comment by fredlcore** [comment]
Has anyone checked if 2.0.20 is also still affected by this?

**Comment by parautenbach** [comment]
> Has anyone checked if 2.0.20 is also still affected by this?

Have you?

**Comment by fredlcore** [comment]
No, otherwise I wouldn't have asked.

**Comment by hjelp** [comment]
2.0.20 still has this issue...

**Comment by fredlcore** [comment]
@ralight, don't want to bother you, but could you give us an indication of where this is on your priority list? I haven't tested version 2.0.22 and also not the 2.1-test version, but is there hope that this will be addressed at some point in the not too distant future? Thanks a lot!

**Comment by ckrey** [comment]
The problem seems to be solved in 2.1 although I don't see an explicit fix for the problem.

The steps to reproduce as described ([see above](https://github.com/eclipse-mosquitto/mosquitto/issues/2887#issuecomment-2195094631)) do not show the problem anymore:

- 100.000 retained messages are stored with QoS1
- SUBSCRIBE with QoS0 receives the expected 100.000 messages, which is **as expected**.


- When SUBSCRIBED to with QoS1 are about `max_queued_messages` and there is the log message `Outgoing messages are being dropped` **as expected**.

@BubuOT and others, it might be worth a try to see if your problems with previous versions have disappeared too.

**Comment by pat1** [comment]
@ckrey 
> The problem seems to be solved in 2.1 although I don't see an explicit fix for the problem.

se at #2526
